### PR TITLE
Allow fsspec/zarr/mfdataset

### PR DESCRIPTION
**Pull request by martindurant**
Requires https://github.com/zarr-developers/zarr-python/pull/606

 - [ ] ~Closes #xxxx~
 - [x] Tests added
 - [x] Passes `isort . && black . && mypy . && flake8`
 - [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
 - [x] New functions/methods are listed in `api.rst`

### CODE REVIEW

File: xarray/backends/api.py [hunk 1]
```diff
@@ -324,6 +324,8 @@ def open_dataset(
     backend_kwargs=None,
     use_cftime=None,
     decode_timedelta=None,
+    storage_options=None,
```
Comment (alexamici) [inline xarray/backends/api.py:327]: Shouldn't `storage_options` and `fs` be `backend_kwargs`?

Comment (martindurant) [reply xarray/backends/api.py:327]: Happy to move it - this is POC, so I just needed to put it somewhere.
However, it's not exactly a part of the backend kwargs, but it just so happens that it's only used by the zarr backend right now.

**Comment by martindurant** [comment]
Question: to eventually get tests to pass, will need changes only just now going into zarr. Those may be released some time soon, but in the meantime is it reasonable to install from master?

**Comment by keewis** [comment]
> is it reasonable to install from master

you might want to take a look at the `upstream-dev` CI which installs `zarr` from github (and is currently passing)

**Comment by dcherian** [comment]
We'll have to maintain backward compatibility with older zarr versions for a bit so you'll have to skip the tests appropriately using a version check

File: doc/io.rst [hunk 2]
```diff
@@ -921,6 +924,22 @@ pass to xarray::
     # read it back
     ds_gcs = xr.open_zarr(gcsmap)
 
+New in v0.16.2: general `fsspec`_ URLs are now parsed and the store set up for you
```
Comment (max-sixty) [inline doc/io.rst:927]: What do you think about swapping the order of these — having a "from v0.16.2" first and then "for older versions..." after (if at all)

File: xarray/backends/api.py [hunk 3]
```diff
@@ -550,11 +557,15 @@ def maybe_decode_store(store, chunks, lock=False):
...
         elif engine == "zarr":
             backend_kwargs = backend_kwargs.copy()
+            backend_kwargs.pop("fs", None)
             overwrite_encoded_chunks = backend_kwargs.pop(
                 "overwrite_encoded_chunks", None
             )
             extra_kwargs["mode"] = "r"
             extra_kwargs["group"] = group
+            if fs is not None:
+                filename_or_obj = fs.get_mapper(filename_or_obj)
```
Comment (alexamici) [inline xarray/backends/api.py:567]: Note that there we are working on refactor of the backend API that among other things aims at removing all knowledge of what backends can or can't do from `open_dataset`. Adding logic inside `if engine == "zarr"` will probably result in merge conflicts.

I would suggest to move the call to `fs.get_mapper(filename_or_obj)` inside the zarr backend.

Comment (martindurant) [reply xarray/backends/api.py:567]: Thanks for the heads up. I already did one slightly complex conflict resolve.

It isn't totally clear, though, that the logic can be buried in the zarr engine for two reasons:
- when using open_mf, the globbing of remote files/directories happens early, before establishing individual zarr instances
- actually the file instances that fsspec makes from URLs can be used by some other backends; that just happens not the be the emphasis here

Happy to go whichever way is most convenient for the library.

File: doc/io.rst (hunk 2 above)
Comment (martindurant) [reply doc/io.rst]: done

File: xarray/backends/api.py [hunk 4]
```diff
@@ -563,6 +565,9 @@ def maybe_decode_store(store, chunks, lock=False):
             )
             extra_kwargs["mode"] = "r"
             extra_kwargs["group"] = group
+            extra_kwargs['storage_options'] = storage_options
+            if fs is not None:
+                filename_or_obj = fs.get_mapper(filename_or_obj)
```
[the remark this answers is not in the document]
Comment (martindurant) [reply xarray/backends/api.py:567]: That works already, and will continue to work. However, the whole point. of this PR is to allow working out those details in a single call to open_dataset, which turns out very convenient for encoding in an Intake catalog, for instance, or indeed for the open_mfdataset implementation in here.

File: doc/io.rst [hunk 5]
```diff
@@ -876,6 +876,7 @@ can be omitted as it will internally be set to ``'a'``.
 
 .. ipython:: python
 
+    ds1 = xr.Dataset(
```
Comment (keewis) [inline doc/io.rst:879]: bad merge? This makes the docs build fail with a `SyntaxError`

Comment (martindurant) [reply doc/io.rst:879]: Hm, interesting. Correcting...

File: doc/io.rst [hunk 6]
```diff
@@ -910,8 +911,28 @@ Cloud Storage Buckets
...
+    ds_gcs = xr.open_dataset(
+        "gcs://<bucket-name>/path.zarr",
+        backend_kwargs={"storage_options": {"project":  '<project-name>', "token": None}},
+        engine="zarr"
+    )
+
+This also works with ``open_mfdataset``, allowing you to pass a list of paths or
+a URL to be interpreted as a glob string.
+
+For older versions, and for writing, you must explicitly set up a ``MutibleMapping``
```
Comment (max-sixty) [inline doc/io.rst:932]:
```suggestion
For older versions, and for writing, you must explicitly set up a ``MutableMapping``
```

**Review by max-sixty — APPROVED** [review]
Thanks @martindurant , this looks good! (I'll wait to see if others have any final thoughts before merging)

**Comment by martindurant** [comment]
(failures look like something in pandas dev)

File: xarray/tests/test_backends.py [hunk 7]
```diff
@@ -4701,3 +4701,25 @@ def test_open_fsspec():
...
+        {
+            "test": xr.DataArray(
+                np.array([0]), dims=("x",), attrs={"scale_factor": 1, "add_offset": 0}
+            )
+        }
+    )
+    ds.to_netcdf(tmp_path / "test.nc")
+    with xr.open_dataset(tmp_path / "test.nc", engine="h5netcdf") as ds2:
+        ds2["test"][0].load()
+>>>>>>> master
```
Comment (keewis) [inline xarray/tests/test_backends.py]:
```suggestion
```

**Comment by keewis** [comment]
> (failures look like something in pandas dev)

yep, that's #4516

**Comment by martindurant** [comment]
One completely unrelated failure (test_polyfit_warnings). Can I please get a final say here (@max-sixty @alexamici ?)

**Comment by rabernat** [comment]
We let this go stale again. I just resolve the conflicts.

File: doc/io.rst [hunk 8]
```diff
@@ -890,8 +922,28 @@ Cloud Storage Buckets
 
 It is possible to read and write xarray datasets directly from / to cloud
 storage buckets using zarr. This example uses the `gcsfs`_ package to provide
-a ``MutableMapping`` interface to `Google Cloud Storage`_, which we can then
-pass to xarray::
+an interface to `Google Cloud Storage`_.
+
+From v0.16.2: general `fsspec`_ URLs are parsed and the store set up for you
+automatically when reading, such that you can open a dataset ina  single
```
Comment (rabernat) [inline doc/io.rst:928]:
```suggestion
automatically when reading, such that you can open a dataset in a single
```

File: xarray/backends/api.py (hunk 3 above)
Comment (rabernat) [reply xarray/backends/api.py:581]: We need to resolve this discussion in order to decide what to do about this PR. Any more thoughts from other devs.

In my view, some of the fsspec logic introduced in the PR should eventually move to the generic `open_mfdataset` function, as it is not really specific to Zarr. However, I don't see a strong downside to adding it to `open_zarr` right now. Eventually `open_zarr` will be deprecated. But the pattern used here could be copied and incorporated into the backend refactor.

**Review by rabernat — APPROVED** [review]
We need to decided what to do with this PR. I have a few comments, but in general I favor merging.

**Comment by dcherian** [comment]
> We need to resolve this discussion in order to decide what to do about this PR. Any more thoughts from other devs.

ping @pydata/xarray

**Comment by martindurant** [comment]
ping again

**Comment by rsignell-usgs** [comment]
I'm really looking forward to getting this merged so I can open the National Water Model Zarr I created last week thusly:  
```python
ds = xr.open_dataset(3://noaa-nwm-retro-v2.0-zarr-pds', engine='zarr', 
        backend_kwargs={'consolidated':True, "storage_options": {'anon':True}})
```
@martindurant tells me this takes only **3 s** with the new async capability!

That would be pretty awesome, because now it takes **1min 15s** to execute thusly:
```python
ds= xr.open_zarr(fsspec.get_mapper('s3://noaa-nwm-retro-v2.0-zarr-pds', anon=True), consolidated=True)
```

**Comment by shoyer** [comment]
We are excited about adding this feature! We love fsspec and think this would be very useful for xarray's users. In the long term, we would love to support fsspec for all the file formats that can handle file objects, e.g., including `engine='h5netcdf'` and `engine='scipy'`.

The concern right now is that this adds special case logic for zarr in open_dataset(), which @alexamici and @aurghs are presently (simultaneously!) trying to remove as part of paying down technical debt in the ongoing backends refactor.

I see two potential paths forwards:

1. Merge this as is. It has good test coverage and porting should (hopefully!) be relatively straightforward to port.
2. Insert this into the new backend API code instead, and require using the v2 backend API instead for this feature.

@alexamici could you please take a look and weigh in here? In particular, it would be helpful if you could point to where this would belong in the new refactor. This is also a good motivation for deleting the "v1" API code as soon as possible in favor of the "v2" code -- nothing is worse than needing to implement a new feature twice!

**Comment by rabernat** [comment]
@rsignell-usgs: note that your example works without this PR (but with the just released zarr 2.6.1) as follows
```python
mapper = fsspec.get_mapper('s3://noaa-nwm-retro-v2.0-zarr-pds')
ds = xr.open_zarr(mapper, consolidated=True)
```

Took 4s on my laptop (outside of AWS).

**Comment by rsignell-usgs** [comment]
@rabernat , awesome!     I was stunned by the difference -- I guess the async loading of coordinate data is the big win, right?

**Comment by rabernat** [comment]
I think @shoyer has laid out the options in a very clear way.

I weakly favor option 2, as I think it preferable in terms of software architecture and our broader roadmap for Xarray. However, I am cognizant of the significant effort that @martinduant has put into this, and I don't want his effort to go to waste.

Some mitigating factors are:
- The example I gave above (https://github.com/pydata/xarray/pull/4461#issuecomment-741939277) shows that one high-impact feature that users want (async capabilities in Zarr) already works, albiet with a different syntax. So this PR is more about convenience.
- Presumably the knowledge about Xarray that Martin has gained by implementing this PR is transferrable to a different context, and so we would not be starting from scratch if we went with 2.

**Comment by martindurant** [comment]
> Martin has gained by implementing this PR is transferrable

I'm not sure, it's been a while now...

**Comment by rafa-guedes** [comment]
> @rabernat , awesome! I was stunned by the difference -- I guess the async loading of coordinate data is the big win, right?

@rsignell-usgs one other thing that can largely speed up loading of metadata / coordinates is ensuring coordinate variables are stored in one single chunk. For this particular dataset, chunk size for `time` coordinate is 672 yielding 339 chunks, which can take a while to load from remote stores. If you rewrite `time` coordinate setting `dset.time.encoding["chunks"] = 227904`, you should see a very large performance increase. One thing we have been doing for the cases of zarr archives that are appended in time, is firstly defining time coordinate with a very large chunk size (e.g., `dset.time.encoding["chunks"] = 10000000`). This ensures time coordinate will still fit in one single chunk after appending over time dimension, and does not affect chunking of the actual data variables.

One thing we have been having performance issues with is with loading zarr archives that have too many chunks (millions), even when metadata is consolidated. There is an [open issue](https://github.com/dask/dask/issues/6363) in dask about this.

**Comment by martindurant** [comment]
All interested parties, please see new attempt at https://github.com/pydata/xarray/pull/4823

**Pull request closed without merging**